driver/src/orderbook/streamed/balance.rs

```diff
@@ -2,63 +2,104 @@ use super::*;
 use error::Error;
 use serde::{Deserialize, Serialize};
 
+/// The balance of a token for a user.
 #[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
 pub struct Balance {
     balance: U256,
-    pub pending_deposit: Option<Flux>,
-    pub pending_withdraw: Option<Flux>,
+    deposit: Flux,
+    withdraw: Flux,
 }
 
 impl Balance {
-    pub fn deposit(&mut self, new_deposit: Flux, current_batch_id: BatchId) {
-        self.update_deposit_balance(current_batch_id);
-        match self.pending_deposit.as_mut() {
-            Some(deposit) => {
-                deposit.amount += new_deposit.amount;
-                deposit.batch_id = new_deposit.batch_id
-            }
-            None => self.pending_deposit = Some(new_deposit),
-        }
+    pub fn deposit(
+        &mut self,
+        amount: U256,
+        deposit_batch_id: BatchId,
+        current_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        self.apply_existing_deposit(current_batch_id)?;
+        // Works like in the smart contract: If there is an existing deposit we override the
+        // batch id and add to the amount. If there is no existing deposit then amount is already 0.
+        let deposit_amount_including_this_deposit = self
+            .deposit
+            .amount
+            .checked_add(amount)
+            .ok_or(Error::MathOverflow)?;
+        self.deposit.batch_id = deposit_batch_id;
+        self.deposit.amount = deposit_amount_including_this_deposit;
+        Ok(())
+    }
+
+    pub fn withdraw_request(&mut self, amount: U256, batch_id: BatchId) {
+        self.withdraw.batch_id = batch_id;
+        self.withdraw.amount = amount;
     }
 
-    pub fn withdraw(&mut self, amount: U256, current_batch_id: BatchId) -> Result<(), Error> {
-        self.update_deposit_balance(current_batch_id);
-        self.pending_withdraw = None;
-        self.balance = self
-            .balance
+    pub fn withdraw(&mut self, amount: U256, batch_id: BatchId) -> Result<(), Error> {
+        if self.withdraw.batch_id >= batch_id {
+            return Err(Error::WithdrawEarlierThanRequested(self.withdraw.batch_id));
+        }
+        if self.withdraw.amount < amount {
+            return Err(Error::WithdrawMoreThanRequested(self.withdraw.amount));
+        }
+        let balance_excluding_withdraw = self.get_balance_internal(batch_id, false)?;
+        let new_balance = balance_excluding_withdraw
             .checked_sub(amount)
-            .ok_or(Error::MathUnderflow)?;
+            .ok_or(Error::WithdrawMoreThanBalance(balance_excluding_withdraw))?;
```

Comment:
Maybe include the amount here as well so that we can see the exact overflow.

Reply:
I have thought about that. My conclusion is that the person adding the event already knows this (because it is in the event) so they would add the event to the context of the error or somethinhg like that if they get an error when applying it. It's the same for `UnknownOrder(OrderId)` where I no longer think the order id makes sense.

Reply:
Good point!